### PR TITLE
Editable module

### PR DESCRIPTION
**Pull request by GreatRSingh**
## Description

Fix #(issue)

Tasks:
- [x] EditableModule
- [x] Example
- [x] Docs

Tests:
  - [x] getparams
  - [x] setparams
  - [x] cachedparamnames
  - [x] getparamnames
  - [x] getuniqueparams
  - [x] setuniqueparams
  - [x] _get_unique_params_idxs
  - [x] assertparams
  - [ ] __assert_method_preserve - [checked with assertparams]
  - [ ] __assert_get_correct_params - [checked with asserparams]
  - [ ] __list_operating_params - [checked with assertparams]
  - [x] _traverse_obj
  - [x] _get_tensors
  - [x] _set_tensors

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+            if grad is None:
+                continue
+            names.append(all_names[i])
+            params.append(all_tensors[i])
+
+        return names, params
+
+
+# traversing functions
+def _traverse_obj(obj, prefix, action, crit, max_depth=20, exception_ids=None):
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:463]: Add type annotations and full docstring

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+                                  action=action,
+                                  crit=crit,
+                                  prefix=name + "." if hasdict else name,
+                                  max_depth=max_depth - 1,
+                                  exception_ids=exception_ids)
+                else:
+                    raise RecursionError("Maximum number of recursion reached")
+
+
+def _get_tensors(obj, prefix="", max_depth=20):
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/editable_module.py:514]: Add type annotations

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 1 above)
Comment (rbharath) [reply deepchem/utils/differentiation_utils/editable_module.py:463]: Add a usage example

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 2 above)
Comment (rbharath) [reply deepchem/utils/differentiation_utils/editable_module.py:514]: Add a usage example

File: deepchem/utils/differentiation_utils/editable_module.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+
+    _traverse_obj(obj,
+                  action=action,
+                  crit=crit,
+                  prefix=prefix,
+                  max_depth=max_depth)
+    return res, names
+
+
+def _set_tensors(obj, all_params, max_depth=20):
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/utils/differentiation_utils/editable_module.py:558]: Add a usage example

**Review by rbharath — COMMENTED** [review]
Requests for improvements to docs and more tests

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 1 above)
Comment (GreatRSingh) [reply deepchem/utils/differentiation_utils/editable_module.py]: Done

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 2 above)
Comment (GreatRSingh) [reply deepchem/utils/differentiation_utils/editable_module.py]: Done

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/utils/differentiation_utils/editable_module.py]: Done

File: deepchem/utils/test/test_differentiation_utils.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            if methodname == "mult":
+                return [prefix + "b"]
+            else:
+                raise KeyError()
+
+    a = torch.tensor(2.0).requires_grad_()
+    x = torch.tensor(0.4).requires_grad_()
+    alpha = A(a)
+    assert alpha.mult(x) == torch.tensor(1.6000)
+    assert alpha.getparamnames("mult") == ['b']
```
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/utils/test/test_differentiation_utils.py:31]: Done

File: deepchem/utils/test/test_differentiation_utils.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        def getparamnames(self, methodname, prefix=""):
+            if methodname == "mult":
+                return [prefix + "b"]
+            else:
+                raise KeyError()
+
+    a = torch.tensor(2.0).requires_grad_()
+    x = torch.tensor(0.4).requires_grad_()
+    alpha = A(a)
+    assert alpha.mult(x) == torch.tensor(1.6000)
```
Comment (rbharath) [inline deepchem/utils/test/test_differentiation_utils.py:30]: Every method in the class should be tested here. For example, setparams, assertparams

**Review by rbharath — COMMENTED** [review]
@GreatRSingh Every method in the new class should have an associated unit test. Please go through and add tests accordingly. Let's do the same for the private methods as well since some of them are complicated

**Comment by rbharath** [comment]
@GreatRSingh Also unit tests are failing. Please fix failing unit tests

**Comment by GreatRSingh** [comment]
> @GreatRSingh Every method in the new class should have an associated unit test. Please go through and add tests accordingly. Let's do the same for the private methods as well since some of them are complicated

I was unable to add tests for private methods directly (as they are private) but they were used in assertparams so they are kind of checked in that.

**Review by rbharath — COMMENTED** [review]
@GreatRSingh Once you fix mypy, should be good to merge

**Review by rbharath — APPROVED** [review]
LGTM

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 1 above)
Comment (enpck100) [reply deepchem/utils/differentiation_utils/editable_module.py]: Validate parameter count mismatches and implement proper cache refresh.

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 2 above)
Comment (enpck100) [reply deepchem/utils/differentiation_utils/editable_module.py]: Improve recursion safety, clarify iterable handling, and validate traversal limits.

File: deepchem/utils/differentiation_utils/editable_module.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (enpck100) [reply deepchem/utils/differentiation_utils/editable_module.py]: Add explicit type annotations and include a clear usage example.

**Review by enpck100 — COMMENTED** [review]
Do this to make it work well

**Comment by enpck100** [comment]
It is working well

File: deepchem/utils/attribute_utils.py [hunk 6]
```diff
@@ -80,7 +80,7 @@ def del_attr(obj: Any, name: str):
 
     Examples
     --------
-    >>> from deepchem.utils.attribute_utils import del_attr
```
Comment (enpck100) [inline deepchem/utils/attribute_utils.py:83]: Public import.

File: deepchem/utils/test/test_attribute_utils.py [hunk 7]
```diff
@@ -17,7 +17,6 @@ def __init__(self):
 
 
 def test_set_attr():
-    from deepchem.utils.attribute_utils import set_attr
```
Comment (enpck100) [inline deepchem/utils/test/test_attribute_utils.py:20]: Direct internal import; should use public API.

File: deepchem/utils/test/test_attribute_utils.py [hunk 8]
```diff
@@ -26,14 +25,13 @@ def __init__(self):
             self.b = 2
 
     obj = MyClass()
-    set_attr(obj, "a", 3)
```
Comment (enpck100) [inline deepchem/utils/test/test_attribute_utils.py:29]: Uses deprecated direct import call.

File: deepchem/utils/test/test_attribute_utils.py [hunk 9]
```diff
@@ -26,14 +25,13 @@ def __init__(self):
...
-    set_attr(obj, "c", 4)
```
Comment (enpck100) [inline deepchem/utils/test/test_attribute_utils.py:30]: Should reference utils.set_attr instead.

File: deepchem/utils/test/test_attribute_utils.py [hunk 10]
```diff
@@ -26,14 +25,13 @@ def __init__(self):
...
+    utils.set_attr(obj, "a", 3)
+    utils.set_attr(obj, "c", 4)
     assert obj.a == 3
     assert obj.c == 4
 
 
 def test_del_attr():
-    from deepchem.utils.attribute_utils import del_attr
```
Comment (enpck100) [inline deepchem/utils/test/test_attribute_utils.py:36]: Avoid internal import; use public namespace.

File: deepchem/utils/test/test_attribute_utils.py [hunk 11]
```diff
@@ -42,7 +40,7 @@ def __init__(self):
             self.b = 2
 
     obj = MyClass()
-    del_attr(obj, "a")
```
Comment (enpck100) [inline deepchem/utils/test/test_attribute_utils.py:45]: Should call utils.del_attr instead.

**Review by enpck100 — COMMENTED** [review]
done

**Pull request merged**